include/deal.II/base/vectorization.h

```diff
@@ -3941,6 +3941,211 @@ operator<<(std::ostream &out, const VectorizedArray<Number, width> &p)
+                       const VectorizedArray<float, 16> &right,
+                       const VectorizedArray<float, 16> &true_values,
+                       const VectorizedArray<float, 16> &false_values)
+{
+  const auto mask = _mm512_cmp_ps(left.data, right.data, predicate);
+  VectorizedArray<float, 16> result;
+  result.data = _mm512_or_ps(_mm512_and_ps(mask, true_values.data),
+                             _mm512_andnot_ps(mask, false_values.data));
+  return result;
+}
+
+
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<double, 8>
+compare_and_apply_mask(const VectorizedArray<double, 8> &left,
+                       const VectorizedArray<double, 8> &right,
+                       const VectorizedArray<double, 8> &true_values,
+                       const VectorizedArray<double, 8> &false_values)
+{
+  const auto mask = _mm512_cmp_pd(left.data, right.data, predicate);
+  VectorizedArray<double, 8> result;
+  result.data = _mm256_or_ps(_mm512_and_ps(mask, true_values.data),
+                             _mm512_andnot_ps(mask, false_values.data));
+  return result;
+}
+
+#  endif
+
+#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 2 && defined(__AVX__)
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<float, 8>
+compare_and_apply_mask(const VectorizedArray<float, 8> &left,
+                       const VectorizedArray<float, 8> &right,
+                       const VectorizedArray<float, 8> &true_values,
+                       const VectorizedArray<float, 8> &false_values)
+{
+  const auto mask = _mm256_cmp_ps(left.data, right.data, predicate);
+
+  VectorizedArray<float, 8> result;
+  result.data = _mm256_or_ps(_mm256_and_ps(mask, true_values.data),
+                             _mm256_andnot_ps(mask, false_values.data));
+  return result;
+}
+
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<double, 4>
+compare_and_apply_mask(const VectorizedArray<double, 4> &left,
+                       const VectorizedArray<double, 4> &right,
+                       const VectorizedArray<double, 4> &true_values,
+                       const VectorizedArray<double, 4> &false_values)
+{
+  const auto mask = _mm256_cmp_pd(left.data, right.data, predicate);
+
+  VectorizedArray<double, 4> result;
+  result.data = _mm256_or_ps(_mm256_and_ps(mask, true_values.data),
+                             _mm256_andnot_ps(mask, false_values.data));
+  return result;
+}
+
+#  endif
+
+#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1 && defined(__SSE2__)
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<float, 4>
+compare_and_apply_mask(const VectorizedArray<float, 4> &left,
+                       const VectorizedArray<float, 4> &right,
+                       const VectorizedArray<float, 4> &true_values,
+                       const VectorizedArray<float, 4> &false_values)
+{
+  const auto mask = _mm_cmp_ps(left.data, right.data, predicate);
```

Comment:
It looks like the `predicate` variable is an AVX thing, so for the SSE part we need the switch statement with the six cases separately, `_mm_cmpeq_ps, _mm_cmplt_ps, _mm_cmple_ps, _mm_cmpgt_ps, _mm_cmpge_ps, _mm_cmpneq_ps`. Same in the double case below.

Reply:
done